rotkehlchen/tasks/historical_balances.py

```diff
@@ -0,0 +1,224 @@
+import logging
+from typing import TYPE_CHECKING, Final
+
+from rotkehlchen.api.websockets.typedefs import ProgressUpdateSubType, WSMessageType
+from rotkehlchen.constants import ZERO
+from rotkehlchen.db.cache import DBCacheStatic
+from rotkehlchen.db.filtering import HistoryEventFilterQuery
+from rotkehlchen.db.history_events import DBHistoryEvents
+from rotkehlchen.fval import FVal
+from rotkehlchen.history.events.structures.types import EventDirection, HistoryEventSubType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import Timestamp
+from rotkehlchen.utils.misc import ts_now
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+    from rotkehlchen.history.events.structures.base import HistoryBaseEntry
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+# Subtypes where asset enters/leaves PROTOCOL bucket (not wallet)
+PROTOCOL_BUCKET_IN_SUBTYPES: Final = (
+    HistoryEventSubType.RECEIVE_WRAPPED,
+    HistoryEventSubType.GENERATE_DEBT,
+)
+PROTOCOL_BUCKET_OUT_SUBTYPES: Final = (
+    HistoryEventSubType.RETURN_WRAPPED,
+    HistoryEventSubType.PAYBACK_DEBT,
+)
+
+
+def get_bucket_for_event(event: 'HistoryBaseEntry') -> tuple[str, str | None, str | None, str]:
+    """Returns (location, location_label, bucket_protocol, asset) bucket.
+
+    Most events affect the wallet bucket (protocol=NULL).
+    Only wrapped token receipts/returns and debt generation/payback
+    affect protocol-specific buckets.
+    """
+    location = event.location.serialize_for_db()
```

Comment:
why do you need to serialize here? Looks expensive since this is called in a loop